benches/bench.rs

```diff
@@ -39,53 +77,135 @@ pub fn jsonrpsee_types_v2(crit: &mut Criterion) {
+		run_round_trip_with_batch(&rt, crit, client, "ws batch requests", Self::REQUEST_TYPE);
+	}
+
+	fn subscriptions(crit: &mut Criterion) {
+		let rt = TokioRuntime::new().unwrap();
+		let url = rt.block_on(helpers::ws_server());
+		let client =
+			Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
+		run_sub_round_trip(&rt, crit, client, "subscriptions");
+	}
 }
 
-pub fn websocket_requests(crit: &mut Criterion) {
-	let rt = TokioRuntime::new().unwrap();
-	let url = rt.block_on(helpers::ws_server());
-	let client =
-		Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
-	run_round_trip(&rt, crit, client.clone(), "ws_round_trip");
-	run_concurrent_round_trip(&rt, crit, client, "ws_concurrent_round_trip");
+pub struct SyncBencher;
+
+impl RequestBencher for SyncBencher {
+	const REQUEST_TYPE: RequestType = RequestType::Sync;
 }
+pub struct AsyncBencher;
 
-pub fn batched_ws_requests(crit: &mut Criterion) {
-	let rt = TokioRuntime::new().unwrap();
-	let url = rt.block_on(helpers::ws_server());
-	let client =
-		Arc::new(rt.block_on(WsClientBuilder::default().max_concurrent_requests(1024 * 1024).build(&url)).unwrap());
-	run_round_trip_with_batch(&rt, crit, client, "ws batch requests");
+impl RequestBencher for AsyncBencher {
+	const REQUEST_TYPE: RequestType = RequestType::Async;
 }
 
-fn run_round_trip(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl Client>, name: &str) {
-	crit.bench_function(name, |b| {
+fn run_round_trip(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl Client>, name: &str, request: RequestType) {
+	crit.bench_function(&request.group_name(name), |b| {
 		b.iter(|| {
 			rt.block_on(async {
-				black_box(client.request::<String>("say_hello", JsonRpcParams::NoParams).await.unwrap());
+				black_box(client.request::<String>(request.method_name(), JsonRpcParams::NoParams).await.unwrap());
 			})
 		})
 	});
 }
 
-/// Benchmark http batch requests over batch sizes of 2, 5, 10, 50 and 100 RPCs in each batch.
-fn run_round_trip_with_batch(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl Client>, name: &str) {
+fn run_sub_round_trip(rt: &TokioRuntime, crit: &mut Criterion, client: Arc<impl SubscriptionClient>, name: &str) {
 	let mut group = crit.benchmark_group(name);
+	group.bench_function("subscribe", |b| {
+		b.iter_with_large_drop(|| {
+			rt.block_on(async {
+				black_box(
+					client
+						.subscribe::<String>(SUB_METHOD_NAME, JsonRpcParams::NoParams, UNSUB_METHOD_NAME)
+						.await
+						.unwrap(),
+				);
+			})
+		})
+	});
+	group.bench_function("subscribe_response", |b| {
+		b.iter_with_setup(
+			|| {
+				rt.block_on(async {
+					client
+						.subscribe::<String>(SUB_METHOD_NAME, JsonRpcParams::NoParams, UNSUB_METHOD_NAME)
+						.await
+						.unwrap()
+				})
+			},
+			|mut sub| {
+				rt.block_on(async { black_box(sub.next().await.unwrap()) });
```

Comment:
ok, and you can't create the subscription outside the bench function because `next` needs &mut self, correct?

Reply:
I can, but in that case, it'd be quite difficult to create a representative benchmark.
`bench` function is called many times, thus if we will try to re-use the subscription made outside of the bench, then the subscription function will have to yield multiple values. If it'd be done without timeouts, it may eat RAM pretty quickly (sender is unbounded). If we will use timeouts, it may affect benchmark results.
In theory, we can try to synchronize somehow received/sent messages using some kind of semaphore, but once again this communication and overhead caused by it will affect the benchmark.

Reply:
fair enough sounds complicated make sense.